File: nodejs-tools/nodejs-compiler-tools/formatter-xml/src/format_xml_prolog_node.test.ts
import { test, expect } from 'vitest';
import { XmlPrologNode } from '@vighnesh153/parser-xml';
import { formatXmlPrologNode } from './format_xml_prolog_node';
import { parseProgram } from './test_utils';

test('should format xml prolog node with 0 indentation level', () => {
  const [parser, program] = parseProgram(`<? xml  encoding =   "utf-8"  version =  "2.0"   ?>`);

  expect(parser.errors.length).toBe(0);
  expect(program.statements.length).toBe(1);
  expect(
    formatXmlPrologNode({
      xmlPrologNode: program.statements[0] as XmlPrologNode,
      indentation: 4,
      indentationLevel: 0,
    })
  ).toMatchInlineSnapshot(`"<?xml encoding="utf-8" version="2.0"?>"`);
});

test('should format xml prolog node with 2 indentation level', () => {
  const [parser, program] = parseProgram(`<? xml  encoding =   "utf-8"  version =  "2.0"   ?>`);

  expect(parser.errors.length).toBe(0);
  expect(program.statements.length).toBe(1);
  expect(
    formatXmlPrologNode({
      xmlPrologNode: program.statements[0] as XmlPrologNode,
      indentation: 4,
      indentationLevel: 2,
    })
  ).toMatchInlineSnapshot(`"        <?xml encoding="utf-8" version="2.0"?>"`);
});

test('should format xml prolog node with 2 indentation level and 3 indentation', () => {
  const [parser, program] = parseProgram(`<? xml  encoding =   "utf-8"  version =  "2.0"   ?>`);

  expect(parser.errors.length).toBe(0);
  expect(program.statements.length).toBe(1);
  expect(
    formatXmlPrologNode({
      xmlPrologNode: program.statements[0] as XmlPrologNode,
      indentation: 3,
      indentationLevel: 2,
    })
  ).toMatchInlineSnapshot(`"      <?xml encoding="utf-8" version="2.0"?>"`);
});

File: tools-jvm/XmlFormatter/src/test/kotlin/dev/vighnesh153/xml/formatter/fmt/FormatXmlPrologNodeTest.kt
package dev.vighnesh153.xml.formatter.fmt

import dev.vighnesh153.xml.formatter.ast.XmlPrologNode
import dev.vighnesh153.xml.formatter.lexer.LexerInputReader
import dev.vighnesh153.xml.formatter.lexer.XmlLexer
import dev.vighnesh153.xml.formatter.parser.XmlParser
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class FormatXmlPrologNodeTest {
    private val lexerInputReader =
        LexerInputReader(
            input = """
                <? xml  encoding =   "utf-8"  version =  "2.0"   ?>
            """
        )
    private val lexer = XmlLexer(lexerInputReader)
    private val parser = XmlParser(lexer)
    private val program = parser.parseProgram()

    @Test
    fun formatsWithZeroIndentation() {
        assertEquals(0, parser.errors.size)
        assertEquals(1, program.statements.size)
        assertEquals(
            "<?xml encoding=\"utf-8\" version=\"2.0\"?>",
            (program.statements.first() as XmlPrologNode).format(
                indentationLevel = 0,
                indentation = 4,
            )
        )
    }

    @Test
    fun formatsWithTwoIndentationLevel() {
        assertEquals(0, parser.errors.size)
        assertEquals(1, program.statements.size)
        assertEquals(
            "        <?xml encoding=\"utf-8\" version=\"2.0\"?>",
            (program.statements.first() as XmlPrologNode).format(
                indentationLevel = 2,
                indentation = 4,
            )
        )
    }

    @Test
    fun formatsWithTwoIndentationLevelAndThreeIndentation() {
        assertEquals(0, parser.errors.size)
        assertEquals(1, program.statements.size)
        assertEquals(
            "      <?xml encoding=\"utf-8\" version=\"2.0\"?>",
            (program.statements.first() as XmlPrologNode).format(
                indentationLevel = 2,
                indentation = 3,
            )
        )
    }
}